### DISCUSSION TITLE
How should I be updating my Tasks for hera v5?

### DISCUSSION THREAD

**Discussion by dmerrick** [argoproj-labs/hera#615] [Q&A] [answered]
Right now I'm using something like this:

```python
# create a simple function so we can use a named Parameter
def run_command(cmd: str) -> str:
  return cmd

task = Task(
  name="cmd-template",
  source=run_command,
  command=["bash"],
  image=docker_image,
  inputs=[Parameter(name="cmd", default="echo Hello Argo")],
)
```

on hera v5, I get the following error:

```
>   ???
E   pydantic.error_wrappers.ValidationError: 1 validation error for Task
E   __root__
E     exactly one of ['template', 'template_ref', 'inline'] must be present (type=value_error)
```

What is the recommended way to update my code to work with the new hera?

(Related to https://github.com/argoproj-labs/hera/discussions/562)

**Comment by flaviuvadan** [comment, marked as answer]
Hey @dmerrick! Thanks for the question! If you want to maintain a structure that closely matches the example you can write something like:
```python
from hera.workflows import Script, Workflow

def run_command(cmd: str) -> str:
  return cmd

task = Script(
  name="cmd-template",
  source=run_command,
  command=["bash"],
  image=docker_image,
  inputs=[Parameter(name="cmd", default="echo Hello Argo")],
)
```

If you want to take advantage of some of the new v5 features ([single script example](https://github.com/argoproj-labs/hera/blob/main/examples/workflows/upstream/retry_script.py) and [Hera runner example](https://github.com/argoproj-labs/hera/blob/main/examples/workflows/callable_coinflip.py)):
```python
from hera.workflows import script, Workflow

@script(image="docker_image", constructor="runner")  # you can also use GlobalHooks to set this! See example above 
def run_command(cmd: str = "echo Hello Argo") -> str:  # the kwarg will auto-create the Parameter with name + default
  return cmd  # if we want to `return` here, need the Hera runner

with Workflow(generate_name="dmerrick-", entrypoint="run_command") as w:
    run_command()
```

## Why

In v4 the idea of a "task" and a "template" were one and the same. The intent was to save users from having to define a template and then use that template in a task. However, that caused confusion.. so they are unwrapped in V5, as they are in Argo Workflows officially. The `@script` decorator and `Script` object are now the familiar script representation of Argo Workflows, which can be used as a task or a step in a workflow. This script can run on its own, as the only template, as in the example above, or as part of a DAG or steps, which Hera offers multiple examples of!

**Comment by dmerrick** [comment]
@flaviuvadan thanks for this detailed answer

We're not using the contextmanager pattern right now, we've been using `workflow.add_task()`. Is there a similar way to add Scripts, or will we need to update everything to use the more-modern hera syntax?
